### PR TITLE
GP Covariance Function Type Hints

### PR DESCRIPTION
**Pull request by jahall**
Adding type-hints to `cov.py`. Also moving `Periodic` to appear below the other stationary kernels.

## New features
- Type hints in `cov.py`

----
:books: Documentation preview :books:: https://pymc--6740.org.readthedocs.build/en/6740/

### CODE REVIEW

File: pymc/gp/cov.py
Comment (bwengals) [inline pymc/gp/cov.py:1075]: args got lost here

File: pymc/gp/cov.py
Comment (bwengals) [inline pymc/gp/cov.py:919]: should be `self.lfunc = handle_args(lengthscale_func, args)`

File: pymc/gp/cov.py
Comment (bwengals) [inline pymc/gp/cov.py]: omega is `TensorLike`

File: pymc/gp/cov.py
Comment (bwengals) [inline pymc/gp/cov.py:88]: might be an ignorant question, but why is the return type a string instead of `Add`?

File: .pre-commit-config.yaml
Comment (bwengals) [inline .pre-commit-config.yaml]: what's this for?

File: pymc/gp/cov.py
Comment (jahall) [reply pymc/gp/cov.py:1075]: It's actually not necessary to pass them when constructing the callable. They're only required when calling the returned func.

File: pymc/gp/cov.py
Comment (jahall) [reply pymc/gp/cov.py]: Done

File: pymc/gp/cov.py
Comment (jahall) [reply pymc/gp/cov.py:88]: Basically because the `Add` class has not yet been defined (as it is a subclass of this one), therefore, as per [PEP 484](https://peps.python.org/pep-0484/#forward-references) you can define it as a string instead and mypy will work accordingly. Though I *think* as of 3.10 you should be able to specify without the string.

File: .pre-commit-config.yaml
Comment (jahall) [reply .pre-commit-config.yaml]: When I ran the pre-commit hooks the pydocstyle was failing for me, complaining that `tomli` was not installed...

File: pymc/gp/cov.py
Comment (bwengals) [reply pymc/gp/cov.py:1075]: oh yeah youre right, ok let's keep

File: .pre-commit-config.yaml
Comment (bwengals) [reply .pre-commit-config.yaml]: Might be better to leave out of this particular PR then... Maybe it would be better raised as an issue?  I just don't know much about precommit and these things and so would rather not slip this in with a GP PR.

Comment (jahall) [reply .pre-commit-config.yaml]: Will remove

**Comment by bwengals** [comment]
Hey @jahall sorry for the delay!  I (finally) enabled the test suite for you.  Looks like `test_handleargs` needs a small update to reflect your change, and there are a couple mypy issues.  There's a script in the repo to run the same mypy checks locally that the test runner does,  `python scripts/run_mypy.py --verbose`.  By the way, thanks for your contributions here and for the new periodic kernel.  Looking forward to using it!

**Comment by jahall** [comment]
@bwengals No problem! That should be mypy/tests fixed now. I'll wait till we merge this PR then will update the WrappedPeriodic PR to reflect these changes.

**Comment by jahall** [comment]
@bwengals ...bit confused on the test failures ...

**Comment by ricardoV94** [comment]
They showed up in another PR, I think it's unrelated, probably something that changed with the latest PyTensor release

**Comment by ricardoV94** [comment]
Should be fixed by https://github.com/pymc-devs/pymc/pull/6782

**Comment by jahall** [comment]
@bwengals @ricardoV94 This should be good to go now after merging Ricardo's fix

**Comment by jahall** [comment]
@bwengals @ricardoV94 Should be good to go now after merging latest main.

**Comment by jahall** [comment]
@bwengals @ricardoV94 Thoughts?

File: pymc/gp/cov.py
Comment (bwengals) [inline pymc/gp/cov.py:148]: Is this factored out for mypys sake?

**Review by bwengals — APPROVED** [review]
thank you @jahall !

File: pymc/gp/cov.py
Comment (jahall) [reply pymc/gp/cov.py:148]: Ah, yes, this was for mypys sake...I think because `pt.alloc` can technically return things other than `TensorVariable`...

**Comment by jahall** [comment]
@ricardoV94 Can we merge this?

File: pymc/gp/cov.py
Comment (ricardoV94) [reply pymc/gp/cov.py:148]: It cannot, but mypy doesn't understand that. If you would like ta take a look at the PyTensor level to see if it can be fixed that would be great!

Comment (jahall) [reply pymc/gp/cov.py:148]: Sure - let me take a look

**Pull request merged**
